### PR TITLE
Gloo support for MacOS: Second try

### PR DESCRIPTION
**Pull request by heiner**
This works for me(tm) on

```
$ uname -a
Darwin joeliel.local 23.4.0 Darwin Kernel Version 23.4.0: Fri Mar 15 00:12:37 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6031 arm64
```

I build the wheel via

```
python build/build.py --bazel_options=--override_repository=xla=~/src/xla/ --noenable_cuda
```

(in JAX, commit `833560deb1a94971c3f607a31b18d493b116e2d0`).

I then install the `jaxlib` wheel and update JAX via

```
uv pip install dist/jaxlib-0.4.32.dev20240812-cp310-cp310-macosx_11_0_arm64.whl --force-reinstall
uv pip install .
```

The following works:

```
$ python -c "import jax; print(jax.default_backend()); print(jax.devices()); print(len(jax.devices()))"
cpu
[CpuDevice(id=0)]
1
```

Likewise with this JAX wheel my use case for `jax_cpu_collectives_implementation:gloo` works on MacOS.

### CODE REVIEW

**Comment by heiner** [comment]
However, when I try to run the test if fails for some seemingly unrelated reason:

```
$ ~/src/jax/build/bazel-6.5.0-darwin-arm64 run --test_output=errors --spawn_strategy=sandboxed //xla/pjrt/cpu:gloo_collectives_test
...
INFO: Running command line: external/bazel_tools/tools/test/test-setup.sh xla/pjrt/cpu/gloo_collectives_test
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //xla/pjrt/cpu:gloo_collectives_test
-----------------------------------------------------------------------------
dyld[43777]: symbol not found in flat namespace '__ZN9benchmark16PrintDefaultHelpEv'
external/bazel_tools/tools/test/test-setup.sh: line 333: 43777 Abort trap: 6           "${TEST_PATH}" "$@" 2>&1
     43778 Done                    | tee -a "${XML_OUTPUT_FILE}.log"
zsh: abort      ~/src/jax/build/bazel-6.5.0-darwin-arm64 run --test_output=errors
```

This seems something related to [google/benchmark](https://github.com/google/benchmark) but I'm not sure how to resolve it.

**Comment by heiner** [comment]
Likewise, if I sync to current `main` the compilation fails due to an unrelated issue.

**Comment by heiner** [comment]
Update: I fixed the test by adding `linkstatic=True`. It now runs like this:

```
$ ~/src/jax/build/bazel-6.5.0-darwin-arm64 run --test_output=errors --spawn_strategy=sandboxed //xla/pjrt/cpu:gloo_collectives_test
Executing tests from //xla/pjrt/cpu:gloo_collectives_test
-----------------------------------------------------------------------------
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from GlooCollectives
[ RUN      ] GlooCollectives.AllReduce
[       OK ] GlooCollectives.AllReduce (2 ms)
[----------] 1 test from GlooCollectives (2 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (2 ms total)
[  PASSED  ] 1 test.
external/bazel_tools/tools/test/test-setup.sh: line 353: 52357 Killed: 9               sleep 10
```

Not sure what the `Killed 9: sleep 10` is about.

File: xla/pjrt/cpu/gloo_collectives_test.cc [hunk 1]
```diff
@@ -25,8 +25,12 @@ limitations under the License.
 #include "absl/status/statusor.h"
 #include "absl/time/time.h"
 #include "absl/types/span.h"
+#ifdef __linux__
 #include "gloo/transport/tcp/attr.h"
 #include "gloo/transport/tcp/device.h"
+#elif __APPLE__
```
Comment (penpornk) [inline xla/pjrt/cpu/gloo_collectives_test.cc:31]: Should this be `#elifdef`? Other places in XLA also check if it's defined.
```suggestion
#elifdef __APPLE__
```

File: xla/python/xla.cc [hunk 2]
```diff
@@ -68,6 +68,10 @@ limitations under the License.
 #include "gloo/transport/tcp/device.h"
 #include "xla/pjrt/cpu/gloo_collectives.h"
 #include "xla/pjrt/cpu/gloo_kv_store.h"
+#elif __APPLE__
```
Comment (penpornk) [inline xla/python/xla.cc:71]: Same question here.
```suggestion
#elifdef __APPLE__
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the fix and the investigation!

**Review by penpornk — APPROVED** [review]

**Pull request closed without merging**